### PR TITLE
Add a "keep" import mode to keep files as-is and export them.

### LINKED ISSUE #38957: Csv files are not included as resources when exporting project
**Godot version:**
3.2.1

**OS/device including version:**
Windows 10 1909

**Issue description:**
https://godotengine.org/qa/71495/exporting-game-with-external-resources-csv-file
I have a project that parses a csv file. When exporting the project, the csv file is not included with the export. A workaround is to change the file extension to .txt (even renaming it to .tres worked when it came to parsing it, though the editor threw parsing errors when starting the project). Renaming it to .txt hides the file in the godot file tree, for whatever reason (I assume it is because it cannot be directly interacted with via the godot ui, but it is very counter intuitive to have a file tree not reflect it's actual contents, but that is a seperate discussion).

**Steps to reproduce:**

1. Open a project
2. Imprt a .csv file
3. Change import preset from "CSV Translation" to "CSV"
4. Open and parse the file content via script
    4a. Observe the file being parsed as expected
5. Export project and include *.csv in the `include_filter` in the `export_presets.cfg` file.
6. Run the exported project
    6a. Observe errors in the console stating something similar to
```
ERROR: File must be opened before use.
At: core/bind/core_bind.cpp:1956
```

**Minimal reproduction project:**
https://github.com/oskarious/godot-csv-problems

### LINKED ISSUE #47061: Can't export non-resource files in directory with .gdignore
**Godot version:**

Does not work in v3.2.4-rc4 or -rc5.
Works in v3.2.4-rc3, v3.2.3, and earlier.

**OS/device including version:**

Using Windows 10 to export either Windows or HTML5.

**Issue description:**

Due to #38957, it is necessary to use .gdignore in directories with .csv files used for data. We use this in @ivoyager to read .csv files containing many thousands of astronomical data points. In v3.2.4-rc3, v3.2.3 and earlier, it is possible to export these projects. In v3.2.4-rc4 and -rc5, exports fail because the directories cannot be opened.

**Steps to reproduce:**
Create project with test_directory containing .gdignore and test.csv. Add "*.csv" to Export/Resources/Filters to export non-resource files/folders. Add code:
```
var dir_path := "res://test_directory"

func _ready():
	var dir := Directory.new()
	if dir.open(dir_path) != OK:
		print("Could not open dir: ", dir_path)
		return
	dir.list_dir_begin()
	var file_name := dir.get_next()
	while file_name:
		if !dir.current_is_dir():
			print(file_name)
		file_name = dir.get_next()
```
This will correctly print "test.csv" in editor mode (all versions) and in exports from v3.2.4-rc3*, v3.2.3, and earlier. But in v3.2.4-rc4 and -rc5, both Windows and HTML5 export will print "Could not open dir: res://test_directory".

(*Note: -rc3 had an unrelated issue that breaks HTML5 export; however, Windows export works as stated above.)

**Minimal reproduction project:**

[test-rc5.zip](https://github.com/godotengine/godot/files/6149793/test-rc5.zip)

### PR DESCRIPTION
**Pull request by reduz**
Just choose this mode, and the file will not be imported (but it will be exported as-is).

image

Cherry picking for 3.x should e relatively easy.

Fixes #38957.
Fixes #47061.

### CODE REVIEW

File: editor/editor_file_system.cpp [hunk 1]
```diff
@@ -1710,6 +1722,17 @@ void EditorFileSystem::_reimport_file(const String &p_file, const Map<StringName
 		late_added_files.insert(p_file); //imported files do not call update_file(), but just in case..
 	}
 
+	print_line("reimport, importer: " + importer_name);
```
Comment (akien-mga) [inline editor/editor_file_system.cpp:1725]: That's going to be quite spammy, might be best to remove.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.3. (Via #47300.)

**Comment by golddotasksquestions** [comment]
I don't understand the nomenclature on this at all. Who keeps the file if I chose "Keep File"? And why is there a "No Import" setting on the Import tab? I literally have to click "reimport" to apply and reimport any of the Settings here. Selecting a "No Import" only then to reimport the file makes no sense to me at all.
Seeing this option I would even have more questions and be hesitant to use it: If Godot has auto imported a CSV and it shows up in the FileSystem, if I select "Keep File (No Import)" and then reimport the file, will it disappear from the FileSystem panel? Do I have to go to the Editor settings and enable "Show hidden Files" in order to reimport it again if I changed my mind?

To a godot beginner user (like a game designer who works a lot with spread sheets), non of this would be intuitive or self explanatory. I'm definitely not a beginner any more and I still don't understand what the expected process now is starting with 3.3 to use CSVs for general purpose data, and not for translations.

Right now in Godot 3.2.3 stable, I can use CSV files without fiddling with the Import tab. I would really like this to stay the same.

**Comment by akien-mga** [comment]
@golddotasksquestions Godot imports all files by default for which it has a ResourceImporter. CSV is one such file type as CSV files are imported by default as "CSV Translations" (`.translation` files) which are compressed and optimized translation resources in binary data used by the TranslationServer.

So as of 3.2.3, this is not accurate:

> Right now in Godot 3.2.3 stable, I can use CSV files without fiddling with the Import tab. I would really like this to stay the same.

What actually happens when you add a CSV file to a project to use it, not as translation, but as raw data, is that you get a stream of errors each time you focus the editor:

```
ERROR: import: Error importing CSV translation: 'a' is not a valid locale.
   At: editor/import/resource_importer_csv_translation.cpp:104.
ERROR: _reimport_file: Error importing 'res://test.csv'.
   At: editor/editor_file_system.cpp:1801.
```

If that doesn't happen for you, it means that you *did* fiddle with the Import tab to select the "CSV" importer instead of "CSV Translation". That importer was a hack added to prevent trying to import the CSV as translations (but instead it would import them as nothing - leading to various bugs, such as `.csv` files being missing once the project is exported, since the import system considers the source files unnecessary given that it will only rely on the imported file).

This PR solves this issue by adding a proper "Keep" mode which tells the importer, on a file by file basis, that this file should not be imported by Godot, i.e. it's meant to be used "manually" from your script with the `File` API.

I appreciate that it might seem confusing to have to go to the "Import" tab to configure the non-import of a given file, but this dock is where the import behavior is configured, and disabling it for a file is part of this configuration.

**Comment by golddotasksquestions** [comment]
Thank you for taking the time to explain this in such detail Akien, also thanks to reduz for improving this issue, I really do appreciate the better CSV support a lot (I'm sorry I have not said this in my previous comment)

I am pretty certain I never fiddled with CSV import settings. I clearly remember last year after up grading my experience was that it "just worked". That's why I left this comment [here](https://github.com/godotengine/godot/issues/38245#issuecomment-721363578).

I think the reason for this is because I never imported the CSV in the first place. I always loaded it from an externally file:
````
func parse_csv():
	var file = File.new()
	file.open(str(os_dependent_csv_path,"data.csv"), file.READ)
	while !file.eof_reached():
		var csv_rows = file.get_csv_line("\t")
		csv.append(csv_rows)

	file.close()
	csv.pop_back() #remove last empty array get_csv_line() has created 
	headers = Array(csv[0])
	cardnumber_idx = headers.find("cardnumber")
	name_idx = headers.find("name")
	type_idx = headers.find("type")
	csv_noheaders = csv 
	csv_noheaders.pop_front() #remove first array (headers) from the csv
````

So I guess this totally is a misunderstanding on my part, I am sorry for the confusion.

---------------------------------------------

Having that cleared up, I still think the naming here does not make any sense.
Personally I would really prefer if CSVs would not be imported as translations by default with Godot 4.0 going forward. 
CSVs have a much broader use in game design than just translation files. Translations seem like a specific usecase to me, I would therefore prefer to have to option to "reimport as translation" and maybe "reimport as text data" (with thext data being the default), rather than "reimport as No Import" or "reimport as Keep File".

**Pull request merged**